### PR TITLE
Decouple `serde` from its `derive` crate

### PR DESCRIPTION
**Pull request by CryZe**
By not activating the `derive` feature on `serde`, the compilation speed can be improved by a lot. This is because `serde` can then compile in parallel to `serde_derive`, allowing it to finish compilation possibly even before `serde_derive`, unblocking all the crates waiting for `serde` to start compiling much sooner.

As it turns out the main deciding factor for how long the compile time of a project is, is primarly determined by the depth of dependencies rather than the width. In other words, a crate's compile times aren't affected by how many crates it depends on, but rather by the longest chain of dependencies that it needs to wait on. In many cases `serde` is part of that long chain, as it is part of a long chain if the `derive` feature is active:

`proc-macro2` compile build script > `proc-macro2` run build script > `proc-macro2` > `quote` > `syn` > `serde_derive` > `serde` > `serde_json` (or any crate that depends on serde)

By decoupling it from `serde_derive`, the chain is shortened and compile times get much better.

Check this issue for a deeper elaboration:
https://github.com/serde-rs/serde/issues/2584

For `wasmtime` I'm seeing a reduction from 24.75s to 22.45s when compiling in `release` mode. This is because wasmtime through `gimli` has a dependency on `indexmap` which can only start compiling when `serde` is finished, which you want to happen as early as possible so some of wasmtime's dependencies can start compiling.

To measure the full effect, the dependencies can't by themselves activate the `derive` feature. I've upstreamed a patch for `fxprof-processed-profile` which was the only dependency that activated it for `wasmtime` (not yet published to crates.io). `wasmtime-cli` and co. may need patches for their dependencies to see a similar improvement.

### CODE REVIEW

File: Cargo.lock
Comment (fitzgen) [inline Cargo.lock:124]: Can you avoid this version bump in this PR? IIUC, decoupling the dependencies is orthogonal from upgrading them, and upgrading dependencies requires a core team member do an audit, which will slow things down here.

**Review by fitzgen — COMMENTED** [review]
Thanks! One comment below.

**Comment by pchickey** [comment]
https://github.com/bytecodealliance/wasmtime/pull/6921 provides the `cargo vet` imports for this PR, which has to be done by core contributors. Please merge with main once that PR has landed.

**Comment by fitzgen** [comment]
That PR merged, so this can be rebased now.

**Review by fitzgen — APPROVED** [review]
Thanks!

**Pull request merged**
